components/raft_log_engine/src/engine.rs

```diff
@@ -89,74 +97,136 @@ impl<W: Seek + Write> Write for ManagedWriter<W> {
+        match self.inner.as_ref() {
+            Either::Left(writer) => writer.truncate(offset),
+            Either::Right(writer) => writer.inner().truncate(offset),
+        }
+    }
+
+    fn sync(&self) -> IoResult<()> {
+        match self.inner.as_ref() {
+            Either::Left(writer) => writer.sync(),
+            Either::Right(writer) => writer.inner().sync(),
+        }
+    }
+
+    fn allocate(&self, offset: usize, size: usize) -> IoResult<()> {
+        match self.inner.as_ref() {
+            Either::Left(writer) => writer.allocate(offset, size),
+            Either::Right(writer) => writer.inner().allocate(offset, size),
+        }
+    }
+}
+
+struct ManagedFileSystem {
+    base_level_file_system: DefaultFileSystem,
     key_manager: Option<Arc<DataKeyManager>>,
     rate_limiter: Option<Arc<IORateLimiter>>,
 }
 
-impl ManagedFileBuilder {
+impl ManagedFileSystem {
     fn new(
         key_manager: Option<Arc<DataKeyManager>>,
         rate_limiter: Option<Arc<IORateLimiter>>,
     ) -> Self {
         Self {
+            base_level_file_system: DefaultFileSystem,
             key_manager,
             rate_limiter,
         }
     }
 }
 
-impl FileBuilder for ManagedFileBuilder {
-    type Reader<R: Seek + Read + Send> = ManagedReader<R>;
-    type Writer<W: Seek + Write + Send> = ManagedWriter<W>;
+struct ManagedHandle {
+    path: PathBuf,
+    base: Arc<<DefaultFileSystem as FileSystem>::Handle>,
+}
+
+impl Handle for ManagedHandle {
+    fn truncate(&self, offset: usize) -> IoResult<()> {
+        self.base.truncate(offset)
+    }
+
+    fn file_size(&self) -> IoResult<usize> {
+        self.base.file_size()
+    }
+}
+
+impl FileSystem for ManagedFileSystem {
+    type Handle = ManagedHandle;
+    type Reader = ManagedReader;
+    type Writer = ManagedWriter;
 
-    fn build_reader<R>(&self, path: &Path, reader: R) -> IoResult<Self::Reader<R>>
-    where
-        R: Seek + Read + Send,
-    {
+    fn create<P: AsRef<Path>>(&self, path: P) -> IoResult<Self::Handle> {
+        Ok(ManagedHandle {
+            path: path.as_ref().to_path_buf(),
+            base: Arc::new(self.base_level_file_system.create(path.as_ref())?),
+        })
+    }
+
+    fn open<P: AsRef<Path>>(&self, path: P) -> IoResult<Self::Handle> {
+        Ok(ManagedHandle {
+            path: path.as_ref().to_path_buf(),
+            base: Arc::new(self.base_level_file_system.open(path.as_ref())?),
+        })
+    }
+
+    fn new_reader(&self, handle: Arc<Self::Handle>) -> IoResult<Self::Reader> {
+        let base_reader = self
+            .base_level_file_system
+            .new_reader(handle.base.clone())?;
         if let Some(ref key_manager) = self.key_manager {
             Ok(ManagedReader {
-                inner: Either::Right(key_manager.open_file_with_reader(path, reader)?),
+                inner: Either::Right(
+                    key_manager.open_file_with_reader(handle.path.clone(), base_reader)?,
```

Comment:
Maybe no need to clone?
```suggestion
                    key_manager.open_file_with_reader(&handle.path, base_reader)?,
```